README.md
# D3 JavaScript Network Graphs from R

Version 0.2.13
[![CRAN Version](http://www.r-pkg.org/badges/version/networkD3)](http://cran.r-project.org/package=networkD3)
[![Build Status](https://travis-ci.org/christophergandrud/networkD3.svg?branch=master)](https://travis-ci.org/christophergandrud/networkD3)
![CRAN Monthly Downloads](http://cranlogs.r-pkg.org/badges/last-month/networkD3)
![CRAN Total Downloads](http://cranlogs.r-pkg.org/badges/grand-total/networkD3)

This README includes information on set up and a number of basic examples.
For more information see the package's [main page](http://christophergandrud.github.io/networkD3/).

## Usage

Here's an example of `simpleNetwork`:

```R
# Create fake data
src <- c("A", "A", "A", "A", "B", "B", "C", "C", "D")
target <- c("B", "C", "D", "J", "E", "F", "G", "H", "I")
networkData <- data.frame(src, target)

# Plot
simpleNetwork(networkData)
```

Here's `forceNetwork`:

```R
# Load data
data(MisLinks)
data(MisNodes)

# Plot
forceNetwork(Links = MisLinks, Nodes = MisNodes, Source = "source",
             Target = "target", Value = "value", NodeID = "name",
             Group = "group", opacity = 0.4,
             colourScale = "d3.scale.category20b()")
```

Here's `sankeyNetwork` using a downloaded JSON data file:

```R
# Recreate Bostock Sankey diagram: http://bost.ocks.org/mike/sankey/
# Load energy projection data
URL <- paste0("https://cdn.rawgit.com/christophergandrud/networkD3/",
              "master/JSONdata/energy.json")
Energy <- jsonlite::fromJSON(URL)

# Plot
sankeyNetwork(Links = Energy$links, Nodes = Energy$nodes, Source = "source",
             Target = "target", Value = "value", NodeID = "name",
             units = "TWh", fontSize = 12, nodeWidth = 30)
```
### Interacting with igraph

You can use [igraph](http://igraph.org/r/) to create network graph data that can be plotted with **networkD3**. The `igraph_to_networkD3` function converts igraph graphs to lists that work well with **networkD3**. For example:

```R
# Load igraph
library(igraph)

# Use igraph to make the graph and find membership
karate <- make_graph("Zachary")
wc <- cluster_walktrap(karate)
members <- membership(wc)

# Convert to object suitable for networkD3
karate_d3 <- igraph_to_networkD3(karate, group = members)

# Create force directed network plot
forceNetwork(Links = karate_d3$links, Nodes = karate_d3$nodes, 
             Source = 'source', Target = 'target', NodeID = 'name', 
             Group = 'group')
```

### Saving to an external file

Use `saveNetwork` to save a network to stand alone HTML file:

```R
library(magrittr)

simpleNetwork(networkData) %>% saveNetwork(file = 'Net1.html')
```

## Note

networkD3 began as a port of
[d3Network](http://christophergandrud.github.io/d3Network/) package to the
[htmlwidgets](https://github.com/ramnathv/htmlwidgets) framework. d3Network is 
no longer supported.


NEWS
# All changes to networkD3 are documented here.

## 0.2.14

- Explose `sinksRight` option for `sankeyNetwork` so that users can decide
not to have the last nodes moved to the right border. Thanks to Florian 
Breitwieser.

## 0.2.13

- `simpleNetwork`, `forceNetwork`, and `sankeyNetwork` should now work with 
`tbl_df` link and node data frames. Thanks to @mexindian for suggesting.

## 0.2.12

- Add resize to sankeyNetwork for flexdashboard. Thanks to Kenton Russell for
adding.

- Fixed a bug where `Source` and `Target` labels were not properly concatenated
in `sankeyNetwork` tooltips. Thanks to Tuija Sonkkila for reporting.

- Improved `igraph_to_networkD3` so that it now handles more general weight 
names. Thanks to Maurits Evers for reporting.

## 0.2.11

- `forceNetwork`, `simpleNetwork`, and `sankeyNetwork` generate errors if data
does not appear to be zero-indexed. Thanks to Peter Meissner for prompting this
addition.

- Bug fix for `igraph_to_networkD3` so that it now accepts the value attribute.
Thanks to Louis Goddard.

- Bug fix for `diagonalNetwork` when using hierarchical lists with singular
connections. Thanks to @RohdeK. 

- Fix viewbox position when rerendered in shiny. Thanks to @RohdeK.

- Added the `iterations` argument to `sankeyNetwork`, which adjusts the y-axis
positioning. Thanks to @giko45.

## 0.2.10

- Added the function `igraph_to_networkD3` to convert an `igraph` class object
to a list that can be used with networkD3.

- Fixed a bug where `linkColour` was not actually passed to the widget for
`diagonalNetwork` and `radialNetwork`. Thanks to Pierre Formont.

## 0.2.9

- `forceNetwork` now allows you to supply a vector of colours to the
`linkColour`argument. This enables the user to highlight links to specific
nodes. Thanks to Garth Tarr. 

- Minor documentation improvements.

## 0.2.8

- Added `NodeGroup` and `LinkGroup` parameters to `sankeyNetwork` so links
can be colourised. Thanks to Edwin de Jonge.

- `fontFamily` is now applied to legends in `forceNetwork`. Thanks to Casey
Shannon.

- Improved bounding behaviour with `forceNetwork`. Now bounds both the links and
nodes. Thanks to Koba Khitalishvili.

## 0.2.7

- Fixed an issue with sankey viewBox sizing.

## 0.2.6

- `sankeyNetwork` fully supports cycles

- `sankeyNetwork` gets same responsive sizing and better fitting introduced
in 0.2.4 for `diagonalNetwork` and `radialNetwork`.

- `sankeyNetwork` gets same full margin control introduced
in 0.2.4 for `diagonalNetwork` and `radialNetwork`.

## 0.2.5

- Added `chordDiagram` to show directed relationships among entities.

## 0.2.4

- More robust margin argument for `diagonalNetwork` and `radialNetwork` allows
for a single value or specification of `top`, `right`, `bottom`, and `left` by
named `vector` or `list`.

- Responsive sizing using the `viewBox` attribute of `svg` for `diagonalNetwork`
and `radialNetwork` should allow for fitting in the container with no fiddling.

## 0.2.3

- Internal improvements to reduce dependencies: no longer depends on RCurl,
plyr, and rjson.

- Updated examples:

 + Using jsonlite makes JSONtoDF obsolete with the `fromJSON` function.

 + All Github data links now use the CDN link from
  [rawgit](https://rawgit.com/), so the examples should be more inline with
  Github raw policies.

## 0.2.2

- `sankeyNetwork` changes:

    + Removes forced font family in favor of inherited from css or specified
    through the `fontFamily` argument.

    + Fixes issue with backslashes in the label for links.

    + Puts unicode right arrow in the tooltip.
    Adds argument to specify units for the tooltip label.

    + Handles cycles by updating this forked repo: <https://github.com/soxofaan/d3-plugin-captain-sankey>.

    + Makes assumptions that `Source` is column 1 and `Target` is column 2 if
    not provided.

## 0.2.1

- Fixed an issue with `forceNetwork` on Firefox. Thanks to @agoldst.

- Fixed an issue where `forceNetwork` node size would not return to
default after mouseover. Thanks to Pascal Pernot.

## 0.2

- Added `dendroNetwork` to create hierarchical cluster network diagrams
(dendrograms).

- `treeNetwork` is DEPRECATED.

- Placed functionality from `treeNetwork` to `radialNetwork`. The new
`radialNetwork` function has the same functionality, but has been renamed to be
more accurately descriptive and avoid confusion with `dendroNetwork`.

- Added `diagonalNetwork`, which creates tree network diagrams using diagonal
instead of radial nodes.

Thanks to Jonathan Owen.

## 0.1.8

- Added `clickAction` argument to `forceNetwork` to allow the user to
pass a JavaScript expression through to be activated on click of a node.
Thanks to Peter Ellis.

- Added `bounded` argument to `forceNetwork` to allow the user to create a
bounding box for the plot. See http://bl.ocks.org/mbostock/1129492.
Thanks to Charles Sese.

- Added `fontFamily` argument to `forceNetwork`, `simpleNetwork`,
`sankeyNetwork`, and `treeNetwork`. Thanks to Peter Ellis.

- Added `opacityNoHover` argument to `forceNetwork` to set the opacity of node
labels when nodes are not hovered over. Thanks to Peter Ellis.

## 0.1.7

- Include `JS` from htmlwidgets, to make it easier for users to pass arbitrary
JS to more arguments.

- Other internal code and documentation improvements.

## 0.1.6

- `fontSize` used for all functions rather than `fontsize`. Thank you to
@Hunter-Github for spotting this inconsistency.

## 0.1.5

- Minor improvement to treeNetwork documentation.
Thanks to Steven Beaupré and MrFlick.

## 0.1.4

- `forceNetwork` gains three new arguments.

    + `legend` allows you to add a node colour legend.

    + `radiusCalculation` and `Nodesize` allow you to vary node radius by
some values.

Thank you to Charles Sese for these additions.

## 0.1.3

- `zoom` argument added to `simpleNetwork` and `forceNetwork` to allow zooming.
Thanks to @timelyportfolio.

- Updated `treeNetwork` URL.

## 0.1.2.2

- Update link in `forceNetwork` example.

## 0.1.2.1


R/utils.R
#' Create character strings that will be evaluated as JavaScript
#'
#' @param ... character string to evaluate
#'
#' @source A direct import of \code{JS} from Ramnath Vaidyanathan, Yihui Xie,
#' JJ Allaire, Joe Cheng and Kenton Russell (2015). \link{htmlwidgets}: HTML
#' Widgets for R. R package version 0.4.
#'
#' @export

JS <- function (...)
{
    x <- c(...)
    if (is.null(x))
        return()
    if (!is.character(x))
        stop("The arguments for JS() must be a chraracter vector")
    x <- paste(x, collapse = "\n")
    structure(x, class = unique(c("JS_EVAL", oldClass(x))))
}



#' Internal function from Wei Luo to convert a data frame to a JSON array
#'
#' @param dtf a data frame object.
#' @source Function from:
#' \url{http://theweiluo.wordpress.com/2011/09/30/r-to-json-for-d3-js-and-protovis/}
#' @keywords internal
#' @noRd
toJSONarray <- function(dtf){
  clnms <- colnames(dtf)

  name.value <- function(i){
    quote <- '';
    if(class(dtf[, i])!='numeric' && class(dtf[, i])!='integer'){
      quote <- '"';
    }
    paste('"', i, '" : ', quote, dtf[,i], quote, sep='')
  }
  objs <- apply(sapply(clnms, name.value), 1, function(x){paste(x,
                                                          collapse=', ')})
  objs <- paste('{', objs, '}')

  res <- paste('[', paste(objs, collapse=', '), ']')

  return(res)
}


#' Read a text file into a single string
#'
#' @source Code taken directly from Ramnath Vaidyanathan's Slidify
#' \url{https://github.com/ramnathv/slidify}.
#' @param doc path to text document
#' @return string with document contents
#' @keywords internal
#' @noRd
read_file <- function(doc, ...){
  paste(readLines(doc, ...), collapse = '\n')
}


#' Utility function to handle margins
#' @param margin an \code{integer}, a named \code{vector} of integers,
#'    or a named \code{list} of integers specifying the margins
#'    (top, right, bottom, and left)
#'    in \code{px}/\code{pixels} for our htmlwidget.  If only a single
#'    \code{integer} is provided, then the value will be assumed to be
#'    the \code{right} margin.
#' @return named \code{list} with top, right, bottom, left margins
#' @noRd
margin_handler <- function(margin){
  # margin can be either a single value or a list with any of
  #    top, right, bottom, left
  # if margin is a single value, then we will stick
  #    with the original behavior of networkD3 and use it for the right margin
  if(!is.null(margin) && length(margin) == 1 && is.null(names(margin))){
    margin <- list(
      top = NULL,
      right = margin,
      bottom = NULL,
      left = NULL
    )
  } else if(!is.null(margin)){
    # if margin is a named vector then convert to list
    if(!is.list(margin) && !is.null(names(margin))){
      margin <- as.list(margin)
    }
    # if we are here then margin should be a list and
    #   we will use the values supplied with NULL as default
    margin <- modifyList(
      list(top = NULL, right = NULL, bottom = NULL, left = NULL),
      margin
    )
  } else {
    # if margin is null, then make it a list of nulls for each position
    margin <- list(top = NULL, right = NULL, bottom = NULL, left = NULL)
  }
}


#' Function to convert igraph graph to a list suitable for networkD3
#'
#' @param g an \code{igraph} class graph object
#' @param group an object that contains node group values, for example, those
#' created with igraph's \code{\link{membership}} function.
#' @param what a character string specifying what to return. If
#' \code{what = 'links'} or \code{what = 'nodes'} only the links or nodes are
#' returned as data frames, respectively. If \code{what = 'both'} then both
#' data frames will be return in a list.
#'
#' @return A list of link and node data frames or only the link or node data
#' frames.
#'
#' @examples
#' # Load igraph
#' library(igraph)
#'
#' # Use igraph to make the graph and find membership
#' karate <- make_graph("Zachary")
#' wc <- cluster_walktrap(karate)
#' members <- membership(wc)
#'
#' # Convert to object suitable for networkD3
#' karate_d3 <- igraph_to_networkD3(karate, group = members)
#'
#' # Create force directed network plot
#' forceNetwork(Links = karate_d3$links, Nodes = karate_d3$nodes,
#'              Source = 'source', Target = 'target', NodeID = 'name',
#'              Group = 'group')
#'
#' \dontrun{
#' # Example with data from data frame
#' # Load data
#' ## Original data from http://results.ref.ac.uk/DownloadSubmissions/ByUoa/21
#' data('SchoolsJournals')
#'
#' # Convert to igraph
#' SchoolsJournals <- graph.data.frame(SchoolsJournals, directed = FALSE)
#'
#' # Remove duplicate edges
#' SchoolsJournals <- simplify(SchoolsJournals)
#'
#' # Find group membership
#' wt <- cluster_walktrap(SchoolsJournals, steps = 6)
#' members <- membership(wt)
#'
#' # Convert igraph to list for networkD3
#' sj_list <- igraph_to_networkD3(SchoolsJournals, group = members)
#'
#' # Plot as a forceDirected Network
#' forceNetwork(Links = sj_list$links, Nodes = sj_list$nodes, Source = 'source',
#'              Target = 'target', NodeID = 'name', Group = 'group',
#'              zoom = TRUE, linkDistance = 200)
#' }
#'
#' @importFrom igraph V as_data_frame graph.data.frame simplify cluster_walktrap membership
#' @importFrom magrittr %>%
#' @export

igraph_to_networkD3 <- function(g, group, what = 'both') {
    # Sanity check
    if (!('igraph' %in% class(g))) stop('g must be an igraph class object.',
                                      call. = FALSE)
    if (!(what %in% c('both', 'links', 'nodes'))) stop('what must be either "nodes", "links", or "both".',
                                                     call. = FALSE)

    # Extract vertices (nodes)
    temp_nodes <- V(g) %>% as.matrix %>% data.frame
    temp_nodes$name <- row.names(temp_nodes)
    names(temp_nodes) <- c('id', 'name')

    # Convert to base 0 (for JavaScript)
    temp_nodes$id <- temp_nodes$id - 1

    # Nodes for output
    nodes <- temp_nodes$name %>% data.frame %>% setNames('name')
    # Include grouping variable if applicable
    if (!missing(group)) {
      group <- as.matrix(group)
      if (nrow(nodes) != nrow(group)) stop('group must have the same number of rows as the number of nodes in g.',
                                          call. = FALSE)
      nodes <- cbind(nodes, group)
    }
    row.names(nodes) <- NULL

    # Convert links from names to numbers
    links <- as_data_frame(g, what = 'edges')
    links <- merge(links, temp_nodes, by.x = 'from', by.y = 'name')
    links <- merge(links, temp_nodes, by.x = 'to', by.y = 'name')
    if (ncol(links) == 5) {
        links <- links[, c(4:5, 3)] %>%
                      setNames(c('source', 'target', 'value'))
    }
    else {
        links <- links[, c('id.x', 'id.y')] %>% setNames(c('source', 'target'))
    }

    # Output requested object
    if (what == 'both') {
      return(list(links = links, nodes = nodes))
    }
    else if (what == 'links') {
      return(links)
    }
    else if (what == 'nodes') {
      return(nodes)
    }
}

#' Check if data is 0 indexed
#' @keywords internal
#' @noRd

check_zero <- function(Source, Target) {
    if (!is.factor(Source) && !is.factor(Target)) {
        SourceTarget <- c(Source, Target)
        if (is.numeric(SourceTarget) | is.integer(SourceTarget)) {
            if (!(0 %in% SourceTarget))
                warning(
                    'It looks like Source/Target is not zero-indexed. This is required in JavaScript and so your plot may not render.',
                    call. = FALSE)
        }
    }
}

#' Convert dplyr created tbl_df to plain data.frame
#' @noRd

tbl_df_strip <- function(x) {
    if('tbl_df' %in% class(x)) {
        message(paste(deparse(substitute(x)),
                      'is a tbl_df. Converting to a plain data frame.'))
        x <- base::as.data.frame(x)
    }
    return(x)
}
